lib/azuremetadata/azuremetadata.py

```diff
@@ -15,25 +23,73 @@ def get_all(self):
-
-        if self._api_version == '2019-08-15':
-            with open('fixtures/metadata-v2019-08-15-multiple-network-if.json', 'r') as json_file:
-                return json.load(json_file)
-        elif self._api_version == '2017-04-02':
-            with open('fixtures/metadata-v2017-04-02.json', 'r') as json_file:
-                return json.load(json_file)
-        else:
-            raise Exception("Unknown API version fixture exception")
+        return self._make_request(
+            "http://169.254.169.254/metadata/instance?api-version={}".format(quote(self._api_version))
+        )
 
     def get_attested_data(self):
-        # FIXME this is a stub while the CLI UX is being figured out
-
-        if self._api_version == '2019-08-15':
-            with open('fixtures/attested-data-v2019-08-15.json', 'r') as json_file:
-                return json.load(json_file)
-        elif self._api_version == '2018-10-01':
-            with open('fixtures/attested-data-v2018-10-01.json', 'r') as json_file:
-                return json.load(json_file)
-        else:
-            raise Exception("Unknown API version fixture exception (attested data)")
+        return self._make_request(
+            "http://169.254.169.254/metadata/attested/document?api-version={}".format(quote(self._api_version))
+        )
+
+    def get_disk_tag(self, device=None):
+        if not device:
+            device = self._find_root_device()
+
+        if not device:
+            return ''
+
+        try:
+            with open(device, 'rb') as fh:
+                fh.seek(65536)
+                return str(uuid.UUID(bytes_le=fh.read(16)))
+        except OSError as e:
+            print("An error occurred when reading disk tag:", file=sys.stderr)
+            print(str(e), file=sys.stderr)
+            return ''
+
+    @staticmethod
+    def _find_root_device():
+        hex_device_id = os.stat("/").st_dev
+        root_device_id = str(os.major(hex_device_id)) + ":" + str(os.minor(hex_device_id))
+
+        devices = glob.glob("/sys/block/*")
+        for device_path in devices:
+            device = os.path.basename(device_path)
+            partitions = glob.glob("/sys/block/{}/{}*".format(device, device))
+
+            for partition in partitions:
+                with open("{}/dev".format(partition), "r") as fh:
+                    device_id = fh.read().strip()
+
+                if device_id == root_device_id:
+                    return "/dev/{}".format(device)
+
+        return None
+
+    @staticmethod
+    def _make_request(url):
+        tries = 0
+        last_error = None
+        while tries < 5:
+            try:
+                req = urllib.request.Request(url, headers={'Metadata': 'true'})
+                response = urllib.request.urlopen(req, timeout=1)
+
+                data = response.read()
+                if isinstance(data, bytes):
+                    data = data.decode('utf-8')
```

Comment:
Is decoding into utf-8 required here ? what encoding comes from response.read() ? Can we be sure utf-8 conversion is safe ? Sorry I suffered so much from encodings which is why I always ask.

Reply:
I've ran into an issue on SLE12 -- it appears that on Python 3.14 `response.read()` returns bytes. and on SLE15 -- it returns a string.

> what encoding comes from response.read()

Good question. I am actually not sure. I guess in 3.17 on SLE15 it became smarter, and it probably tries to convert to whatever encoding the server has supplied in the respective HTTP header. I'll double check what encoding it has when this magic happens.

> Can we be sure utf-8 conversion is safe ?

I think we can. The response is supposed to be valid JSON, I guess it would be even safe to convert it to ASCII -- all UTF-8 characters are supposed to be escaped with `\u` sequences.

Reply:
The response content type is `Content-Type: application/json; charset=utf-8`. And there's can indeed be UTF-8 encoded data in it:

```
tagsList[0]:
    name: asdasd
    value: ivan-test-1-пыщь-пыщь-( ͡° ͜ʖ ͡°)
```

Decoding it as `utf-8` should be fine, I don't expect the response to have any invalid UTF-8 code points.